Sources/BraintreeVenmo/BTVenmoAppSwitchRequestURL.swift

```diff
@@ -0,0 +1,104 @@
+    ///   - bundleDisplayName: The bundle display name for the current app
+    ///   - environment: The environment, e.g. "production" or "sandbox"
+    ///   - paymentContextID: The Venmo payment context ID (optional)
+    ///   - metadata: Additional Braintree metadata
+    ///   - Returns: The resulting URL, or `nil` if any of the required parameters are `nil`.
+    // TODO: method can be internal once rest of Venmo is in Swift
+    // TODO: consider making these required and doing this check at the call site when BTVenmo_Client is in Swift
+    @objc(appSwitchURLForMerchantID:accessToken:returnURLScheme:bundleDisplayName:environment:paymentContextID:metadata:)
+    public static func appSwitch(
+        forMerchantID merchantID: String?,
+        accessToken: String?,
+        returnURLScheme: String?,
+        bundleDisplayName: String?,
+        environment: String?,
+        paymentContextID: String?,
+        metadata: BTClientMetadata?
+    ) -> URL? {
+        let successReturnURL = returnURL(with: returnURLScheme, result: "success")
+        let errorReturnURL = returnURL(with: returnURLScheme, result: "error")
+        let cancelReturnURL = returnURL(with: returnURLScheme, result: "cancel")
+
+        guard let successReturnURL,
+                let errorReturnURL,
+                let cancelReturnURL,
+                let accessToken,
+                let metadata,
+                let bundleDisplayName,
+                let environment,
+                let merchantID
+        else {
+            return nil
+        }
+
+        let venmoMetadata: [String: String] = [
+            "version": BTCoreConstants.braintreeSDKVersion,
+            "sessionId": metadata.sessionID,
+            "integration": metadata.integrationString,
+            "platform": "ios"
+        ]
+
+        let braintreeData: [String: Any] = ["_meta": venmoMetadata]
+        let serializedBraintreeData = try? JSONSerialization.data(withJSONObject: braintreeData)
+        let base64EncodedBraintreeData = serializedBraintreeData?.base64EncodedString()
+
+        var appSwitchParameters: [String: Any] = [
+            "x-success": successReturnURL,
+            "x-error": errorReturnURL,
+            "x-cancel": cancelReturnURL,
+            "x-source": bundleDisplayName,
+            "braintree_merchant_id": merchantID,
+            "braintree_access_token": accessToken,
+            "braintree_environment": environment,
+            "braintree_sdk_data": base64EncodedBraintreeData ?? ""
+        ]
+
+        if let paymentContextID {
+            appSwitchParameters["resource_id"] = paymentContextID
+        }
+
+        var components = appSwitchBaseURLComponents()
+        components?.percentEncodedQuery = BTURLUtils.queryString(from: appSwitchParameters as NSDictionary)
+        return components?.url
+    }
+
+    // MARK: - Internal Helper Methods
+
+    static func returnURL(with scheme: String?, result: String) -> URL? {
```

Comment:
We can probably update this to require a `scheme` and return non-optional too. I think we could do it now and move the nil check for `scheme` to the `init`, or lump it in with the TODO you have above the `init`.

Reply:
Yeah, updating the scheme to be non-optional makes sense. I'll make this change. For the return, `URLComponents` [url](https://developer.apple.com/documentation/foundation/urlcomponents/1780032-url) is optional so I think we want to keep this as an optional `URL` return.